Implement `ComponentContext` and deserialize more data

StarLederer opened the pull request:
# Motivation

At the moment some components (e.g. `Image`) skip deserializing critical data inside their `fromArchive` method (e.g. image data). This is presumably because it was previously not possible to load the data inside `fromArchive` as it did not have a reference to the Wordprocessing file data (`Archive`) or the [relationships](http://officeopenxml.com/anatomyofOOXML.php) that might be used to map string IDs to paths in the `Archive`. The data that is currently impossible to load, should not be.

# Solution

This PR solves the problem by introducing `ComponentContext`. `ComponentContext` can be though of as an XML element scope, it carries references to things that ancestor nodes might want to make available to their children. The context usually originates in `DocumentXml`, but can be cloned, modified or completely stopped by any component that wishes to do so. That way children see the context the way the nearest ancestor decides. This behavior can be thought of the same as  variable shadowing in JavaScript.

`ComponentContext` is passed to `fromArchive` method of all components, usually by `createChildComponentsFromNodes`; and to `createChildComponentsFromNodes` itself by `DocumentXml` which deserializes the root and instantiates the initial version of the context.

# Additional improvements

Since this PR makes it very easy for various components (e.g. `Image`) to deserialize data that is currently omitted (e.g. image data) missing data deserialization has already been added to `Image` which now loads image data and mime type.

*Please see the commit messages for an exhaustive change log. Note that some changes are not related to the PR but are included because they would be too small to make individual PRs for and do not change the public interface*

wvbe commented on src/classes/Component.ts:
```diff
@@ -1,6 +1,7 @@
 import { type DocumentXml } from '../files/DocumentXml.ts';
 import { type FooterXml, type HeaderXml } from '../files/HeaderFooterXml.ts';
 import { type RelationshipsXml } from '../files/RelationshipsXml.ts';
+import { Archive } from './Archive.ts';
```
`type Archive`

wvbe commented:
@StarLederer thank you for this PR and the motivation to go with it. I see your intent clearly, and your changes so far are beautiful. Really my only worthwhile comment is about the optionality of `ComponentContext`'s properties, and I notice that I've myself been lax with `type` annotations on imports.

Fully intending to merge this, but hoping to do so together with the change to get `Image`'s byte data as that will avoid your beautiful code from being dead for a hot second.

🙏 💯

StarLederer commented:
> my only worthwhile comment is about the optionality of ComponentContext's properties

There are situations (already) in which we cannot provide all of the context details. In fact, I left a @TODO question about being unable to provide `relationships` inside `CommentsXml`. Making context fields optional makes it much easier to deal with situations where the parent cannot/does not want to publish them to their children.

> hoping to (merge this) together with the change to get Image's byte data

Added it and updated PR description. This also introduces a new @TODO that is a question to you about handling an error case but it is minor and is mainly about your preference. I'll convert the PR to draft until we are able to clarify that on Monday

StarLederer commented:
My fork could no longer be merged automatically so I moved the commits to a new branch, reset the develop to match upstream, and rebased the new branch on new develop, however, this automatically closed this PR because head was no longer ahead, I cannot edit head of closed PR, and I cannot reopen it because the current head is not ahead. I opened a new PR with the new head [here](https://github.com/wvbe/docxml/pull/29)
